### PR TITLE
Pre-size hash maps in BalancedShardsAllocator

### PR DESCRIPTION
**Pull request by idegtiarenko**

### CODE REVIEW

**Review by DaveCTurner — APPROVED** [review]
LGTM but it would be good to see some supporting data (e.g. a before/after flamegraph) to quantify the benefit

**Comment by idegtiarenko** [comment]
Without sizing:

without_sizing

With sizing:

with_sizing

Results are fairly similar, however when setting initial size we should have less resizes and less arrays allocated for them (Especially when using this code in a loop during desired balance calculation).

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderTests.java [hunk 1]
```diff
@@ -521,10 +521,10 @@ public void testDiskThresholdWithAbsoluteSizes() {
...
 
         logShardStates(clusterState);
-        // primary shard already has been relocated away
-        assertThat(clusterState.getRoutingNodes().node(nodeWithPrimary).size(), equalTo(0));
-        // node with increased space still has its shard
-        assertThat(clusterState.getRoutingNodes().node(nodeWithoutPrimary).size(), equalTo(1));
+        assertThat(
+            clusterState.getRoutingNodes().node(nodeWithPrimary).size() + clusterState.getRoutingNodes().node(nodeWithoutPrimary).size(),
+            equalTo(1)
+        );
```
Comment (idegtiarenko) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderTests.java:527]: This assertion assumed that primary (allocated on a node with a higher disk usage) is moved before replica (allocated on a node with a lower disk usage). In fact we do not have such guarantee and this is determined y iteration order.

**Review by DaveCTurner — COMMENTED** [review]
Are the changes you've made to `AwarenessAllocationTests` and `DiskThresholdDeciderTests` related to the pre-sizing of the maps or are they pre-existing test bugs? They don't look related so I think I'd rather they were separate PRs unless they need to be here.

**Comment by idegtiarenko** [comment]
>Are the changes you've made to AwarenessAllocationTests and DiskThresholdDeciderTests related to the pre-sizing of the maps or are they pre-existing test bugs? They don't look related so I think I'd rather they were separate PRs unless they need to be here.

Pre-sizing affected hash map iteration order and revealed tests that depended on it:
```
No pre-sizing:
capacity=16
{A-0=Node(A-0), A-1=Node(A-1), A-2=Node(A-2), A-3=Node(A-3), A-4=Node(A-4)}
"A-0" -> {BalancedShardsAllocator$ModelNode@5011} "Node(A-0)"
"A-1" -> {BalancedShardsAllocator$ModelNode@5013} "Node(A-1)"
"A-2" -> {BalancedShardsAllocator$ModelNode@5015} "Node(A-2)"
"A-3" -> {BalancedShardsAllocator$ModelNode@5017} "Node(A-3)"
"A-4" -> {BalancedShardsAllocator$ModelNode@5019} "Node(A-4)"

With pre-sizing
capacity=8
{A-4=Node(A-4), A-0=Node(A-0), A-1=Node(A-1), A-2=Node(A-2), A-3=Node(A-3)}
"A-4" -> {BalancedShardsAllocator$ModelNode@4977} "Node(A-4)"
"A-0" -> {BalancedShardsAllocator$ModelNode@4979} "Node(A-0)"
"A-1" -> {BalancedShardsAllocator$ModelNode@4981} "Node(A-1)"
"A-2" -> {BalancedShardsAllocator$ModelNode@4983} "Node(A-2)"
"A-3" -> {BalancedShardsAllocator$ModelNode@4985} "Node(A-3)"
```

**Comment by DaveCTurner** [comment]
> Pre-sizing affected hash map iteration order and revealed tests that are depending on it:

Gotcha, I see now.

**Pull request merged**